pkg/core/blockchain.go

```diff
@@ -729,6 +953,147 @@ func (bc *Blockchain) tryRunGC(oldHeight uint32) time.Duration {
+	for i := range prefixes {
+		var (
+			acc             util.Uint160
+			trInfo          *state.TokenTransferInfo
+			removeFollowing bool
+			seekErr         error
+		)
+
+		cache.Store.Seek(storage.SeekRange{
+			Prefix:    prefixes[i : i+1],
+			Backwards: false, // From oldest to newest batch.
+		}, func(k, v []byte) bool {
+			var batchAcc util.Uint160
+			copy(batchAcc[:], k[1:])
+
+			if batchAcc != acc { // Some new account we're iterating over.
+				if trInfo != nil {
+					cache.PutTokenTransferInfo(acc, trInfo)
+				}
+				acc = batchAcc
+				trInfo = nil
+				removeFollowing = false
+			} else if removeFollowing {
+				cache.Store.Delete(slice.Copy(k))
+				return seekErr == nil
+			}
+
+			r := io.NewBinReaderFromBuf(v[1:])
+			l := len(v)
+			bytesRead := 1 // 1 is for batch size byte which is read by default.
+			var (
+				oldBatchSize = v[0]
+				newBatchSize byte
+			)
+		traverseBatchLoop:
+			for i := byte(0); i < v[0]; i++ { // From oldest to newest transfer of the batch.
+				var (
+					t       *state.NEP17Transfer
+					isNEP11 bool
+				)
+				switch k[0] {
+				case byte(storage.STNEP11Transfers):
+					tr := new(state.NEP11Transfer)
+					tr.DecodeBinary(r)
+					if r.Err != nil {
+						seekErr = fmt.Errorf("failed to decode subsequent NEP11 transfer: %w", r.Err)
+						break traverseBatchLoop
+					}
+					t = &tr.NEP17Transfer
+					isNEP11 = true
+				case byte(storage.STNEP17Transfers):
+					tr := new(state.NEP17Transfer)
+					tr.DecodeBinary(r)
+					if r.Err != nil {
+						seekErr = fmt.Errorf("failed to decode subsequent NEP17 transfer: %w", r.Err)
+						break traverseBatchLoop
+					}
+					t = tr
+				}
+
+				if t.Block > height {
+					break
+				}
+				bytesRead = l - r.Len() // Including batch size byte.
+				newBatchSize++
+				if trInfo == nil {
+					var err error
+					trInfo, err = cache.GetTokenTransferInfo(batchAcc)
+					if err != nil {
+						seekErr = fmt.Errorf("failed to retrieve token transfer info for %s: %w", batchAcc.StringLE(), r.Err)
+						break traverseBatchLoop
+					}
+				}
+				appendTokenTransferInfo(trInfo, t.Asset, t.Block, t.Timestamp, isNEP11, newBatchSize >= state.TokenTransferBatchSize)
+			}
+			if newBatchSize == oldBatchSize {
+				// The batch is already in storage and doesn't need to be changed.
+				return seekErr == nil
+			}
+			if newBatchSize > 0 {
+				v[0] = newBatchSize
+				cache.Store.Put(k, v[:bytesRead])
+			} else {
+				cache.Store.Delete(k)
+				removeFollowing = true
+			}
+			return seekErr == nil
+		})
+		if seekErr != nil {
+			return seekErr
+		}
+		if trInfo != nil {
+			cache.PutTokenTransferInfo(acc, trInfo)
+		}
+	}
+	return nil
+}
+
+// appendTokenTransferInfo is a helper for resetTransfers that updates token transfer info
+// wrt the given transfer that was added to the subsequent transfer batch.
+func appendTokenTransferInfo(transferData *state.TokenTransferInfo,
```

Comment:
Can `appendTokenTransfer` make some use of it?

Reply:
Technically, it can, but the amount of code won't be reduced because we still need `nextBatch` and `currTimestamp` information to store the subsequent token transfer log.

Reply:
I've added a separate commit for this, but I'd still prefer not to reuse it and leave `appendTokenTransfer` as is.

Reply:
It's OK to me. `TokenTransferInfo` can be simplified a bit eventually.